### ISSUE TITLE
Nested remote catalog example

### ISSUE THREAD

**Issue by gwbischof** [intake/intake#436]
I am having some trouble getting a nested remote catalog working.

Example:
https://github.com/gwbischof/remote_test

This repo attempts to make a nested catalog. It works correctly locally, but fails remotely.
There is a `pytest` that starts an intake-server and tries to access the nested entries, this test fails with:
`TypeError: __init__() got an unexpected keyword argument 'shape'`

Could I have some help to get this working?

The next step is to make the partition argument of `get_partition` more complex than an integer.  What is the right way to do this?

Could someone explain the significance of name, container, container_map, and plugin_registry?

I think this might make a good example for the docs.

Thank you! 
Garrett

**Comment by martindurant** [comment]
There ought to be docstrings on the attrbited of the base Catalog class, perhaps. 

- name: the handle for the class, so you can do `driver: catalog` instead of `driver: intake.catalog.base.Catalog`
- plugin_registry: the mapping from the names in the previous point to the class object. Some catalogs define their own and only get loaded when the cat is loaded.
- container: the basic type of data that the class returns, such as "dataframe"
- container_map: a mapping from names like "dataframe" to sources which allow you to read this type of data from an Intake server. This essentially concerns how to serialise and package the data type. If a data source has a driver whose container isn't in container_map, you'll not be able to read it from a server

> make the partition argument of get_partition more complex than an integer

get_partition takes any serialisable argument, but the argument must make sense for the container type of the source in question, e.g., the "dataframe" type only takes an integer, but "ndarray" type takes a tuple of integers.


I have not yet had a look at your code to see what's going wrong, but I hope the above helps. Please feel free to PR to add doctrings or to the [glossary](https://intake.readthedocs.io/en/latest/glossary.html).

**Comment by danielballan** [comment]
Thanks, @martindurant.


The direct cause of the error

```
TypeError: __init__() got an unexpected keyword argument 'shape'
```

is that `InnerCatalog` inherits the `name` `'catalog'` from its base class `intake.catalog.base.Catalog`, but it has a different `__init__` signature than `intake.catalog.base.Catalog`. Specifically, it adds the required argument `shape`. Intake maps `'catalog'` to `intake.catalog.base.Catalog` and then tries to unpack the `response['args']` into that, in effect:

```py
intake.catalog.base.Catalog(**{'metadata': {'catalog_dir': ''}, 'shape': 'circle'})
```

hence the `TypeError`. I believe are two solutions possible:

1. Give `InnerCatalog` a distinct `name` and corresponding `container`.
2. Forbid `direct_access`, so that intake builds a proxy object instead and does not try to pass `shape` around on the client side.

I have a demonstration of (2) working in 
https://github.com/gwbischof/remote_test/pull/1, and the tests pass.

**Comment by martindurant** [comment]
> Give InnerCatalog a distinct name and corresponding container.

The container can and probably should still be "catalog", but indeed the name should be different, since it refers to this particular class only.

**Comment by gwbischof** [comment]
got it working

**Comment by martindurant** [comment]
If you can publish your finished product, I'm sure the world will be grateful.

**Comment by danielballan** [comment]
Yes, it would be good to generate documentation as an outcome of this issue.

**Comment by gwbischof** [comment]
I understand enough to get it working, but I don't fully understand the intention of all of the pieces. I'll write something up. But my interpretation of the significance of the different parts might differ from the intended significance. Also, I think my explanation will be incomplete because some of the parts are still a mystery to me. For example, how to make caching on the server side behave the way you want.
